Kvm intercept cr3

Wenzel opened the pull request:
Started implementing CR3 interception for KVM.

Wenzel commented:
So, I implemented 2 new APIs:

~~~Rust
    // toggle an event interception
    fn toggle_intercept(
        &mut self,
        _vcpu: u16,
        _event_type: EventType,
        _enabled: bool,
    ) -> Result<(), Box<dyn Error>> {
        unimplemented!();
    }

    // listen and return the next event, or return None
    fn listen(&self, _timeout: u32) -> Result<Option<Event>, Box<dyn Error>> {
        unimplemented!();
    }
~~~

As well as a simple example to listen for the next `100` `CR3` changes.

But it doesn't work as expected:
![Capture d’écran de 2020-01-25 15-15-56](https://user-images.githubusercontent.com/964610/73122480-c49c4f80-3f85-11ea-8da8-b495ea30cfff.png)

I'm getting a segfault ATM, and I think I should have a second look at my `pop_event` implementation:
https://github.com/Wenzel/kvmi/blob/master/src/lib.rs#L215

@kylerky, if you have a few minutes, can you tell me if you see something absurd that i'm doing there, which would lead to a segfault ? (apart from not freeing the kvmi_event afterwards)

Thanks.

kylerky commented:
No idea. I suggest using Valgrind to look for memory corruptions.

I notice that pop_event succeeded when first called. Therefore, it may be that the bug is somewhere else.

Wenzel commented:
@kylerky I found the reason of the crash, I was dereferencing a null pointer, because I wasn't properly checking `kvmi_pop_event` 's return code:
fixed in https://github.com/Wenzel/kvmi/pull/12

The I discovered that the `kvmi` crate was not up to date with `kvmi-v6`, which requires new headers in event replies:
https://github.com/Wenzel/kvmi/pull/13

Now I'm stuck at designing an API to would let me set an event reply once I received it.
I added a new API to the trait:

~~~Rust
fn reply_event(&self, _event: Event, _reply_type: EventReplyType) -> Result<(), Box<dyn Error>> {
        unimplemented!();
}
~~~

Now I need to store the `kvmi_event` structure inside the generic libmicromi's `Event` struct.

But how can I do that ?
I needs to be generic enough so I can store a `XenEvent` or `VirtualBoxEvent`.

This is how I represent my Events right now:
~~~Rust
pub enum EventType {
    Cr { cr_type: CrType, new: u64, old: u64 },
}

pub struct Event {
    pub kind: EventType,
    driver_event: ????? // should be able to store a KVMiEvent, XenEvent or VirtualBoxEvent
}
~~~

Any ideas @kylerky ?

kylerky commented:
```rust
pub struct Event {
    common: EventCommon,
    extra: EventExtra,
}

pub enum EventExtra<CR>
    where CR: CREvent
{
    CR(CR),
    Pause,
}

pub trait CREvent {
    fn get_old() -> u64 {
        ...
    }
    fn get_new() -> u64 {
        ...
    }
}

pub struct KVMCrEvent {
    pub fn construct_reply(...) -> KVMCrReply {

    }
}

impl CREvent for KVMCrEvent {...}

struct KVMCrReply {...}
impl Reply for KVMCrReply {...}

fn main() {
    let kvm_cr_event = ...;
    let reply = kvm_cr_event.construct_reply(Continue, new);
    dom.send(reply);
}

fn send<T: Reply>(reply: T) { ... }

```

kylerky commented:
I think it is hard to fit `reply_event` into a single function because different types of event ask for different kinds of replies. We should at least go for a generic function.

In my proposal above, I go further as to having event as builders for event replies, and then we can use a generic `send` to send them.

Wenzel commented:
@kylerky , @rageagainsthepc

I finished a basic implementation of CR interception on KVM.
It doesn't handle modifying the CR values for now, we can deal with it in a separate PR (I'll create an issue for that and we propose ideas there).

rageagainsthepc commented on src/driver/kvm.rs:
```diff
@@ -106,13 +132,82 @@ impl Introspectable for Kvm {
+    fn listen(&mut self, timeout: u32) -> Result<Option<Event>, Box<dyn Error>> {
+        // wait for next event and pop it
+        debug!("wait for next event");
+        if self.kvmi.wait_event(timeout.try_into().unwrap())?.is_none() {
```
nitpick: `kvmi.wait_event` returns `None` or an empty `Some()`. I would not expect Some() to yield no value. Maybe use a custom enum event/no_event instead?

rageagainsthepc replied:
Also, I don't see why `wait_event` shouldn't return the event directly. Is there a special reason for not popping the event directly in `wait_event`?

Wenzel replied:
I implemented the KVMi Rust API by doing a 1:1 mapping with the libkvmi functions.
But I suppose we could wrap these 2 calls into a single one `pop_event()` function.

rageagainsthepc replied:
Either way works. :+1:

rageagainsthepc commented on src/driver/kvm.rs:
```diff
@@ -106,13 +132,82 @@ impl Introspectable for Kvm {
+                new,
+                old,
+            },
+            KVMiEventType::PauseVCPU => panic!("Unexpected PauseVCPU event. It should have been poped by resume VM. (Did you forgot to resume your VM ?)"),
```
nitpick: *popped, *forget

rageagainsthepc commented on src/driver/kvm.rs:
```diff
@@ -1,32 +1,58 @@
     kvmi: KVMi,
     expect_pause_ev: u32,
+    // VCPU -> KVMiEvent
+    map_events: HashMap<u16, KVMiEvent>,
```
nitpick: From what I understand you want to register a single event per vcpu, right? Isn't a hash map a little bit of an overkill for this purpose? I would just use a vector instead. You don't necessarily need to change this, I mainly want to know if I have missed something. If so, please tell me ;)

Wenzel replied:
Yes, it's totally overkill.

I just wanted to store the KVMEvent in the KVM driver, and associate its VCPU id to find it back for the response.

A vector is good also.
 
And i'm not storing singlestep events but general KVMevents.

Wenzel replied:
The last version uses a Vector instead.

rageagainsthepc commented on src/api.rs:
```diff
@@ -71,7 +76,63 @@ pub trait Introspectable {
+
+    fn reply_event(
+        &mut self,
+        _event: &Event,
```
Just a thought: It would make sense to consume the event here, right? This would automatically make sure that the user replies only once per event. Now we just need to ensure that the user sends a reply at all. ;)

rageagainsthepc commented on examples/cr3-events.rs:
```diff
@@ -0,0 +1,84 @@
+
+    let ev_per_sec = i as f64 / duration.as_secs_f64();
+    println!(
+        "Catched {} events in {:.2} seconds ({:.2} events/sec)",
```
nitpick: *Caught

Wenzel commented:
@rageagainsthepc merging this one.
